File: frontend/src/types/project.ts
import type { Task } from "./task";

export type ProjectStatus = "planning" | "active" | "completed" | "on_hold";

export interface Project {
  id: string;
  name: string;
  description: string;
  budget: number;
  status: ProjectStatus;
  start_date: string;
  end_date: string;
  tasks: Task[];
  actual_cost: number;
  progress: number;
  created_at: string;
  updated_at: string;
}

export type ProjectCreate = Omit<
  Project,
  "id" | "tasks" | "actual_cost" | "progress" | "created_at" | "updated_at"
>;
export type ProjectUpdate = Partial<ProjectCreate>;

export interface ProjectListParams {
  page?: number;
  per_page?: number;
  status?: ProjectStatus;
}

export interface ProjectCost {
  project_id: string;
  project_name: string;
  budget: number;
  actual_cost: number;
  remaining_budget: number;
  task_costs: {
    task_id: string;
    task_title: string;
    assignee_name: string;
    man_days: number;
    daily_rate: number;
    cost: number;
  }[];
}

export interface BudgetVsActualData {
  project_id: string;
  project_name: string;
  budget: number;
  actual_cost: number;
  budget_consumption_rate: number;
  remaining_budget: number;
  status: string;
}

File: backend/app/schemas/project.py
from datetime import date, datetime
from typing import List, Literal, Optional

from pydantic import BaseModel, Field, field_validator


class ProjectCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    description: str = Field(default="", max_length=500)
    budget: float = Field(..., ge=0.1, le=99999.9)
    status: Literal["planning", "active", "completed", "on_hold"] = "planning"
    start_date: date
    end_date: date

    @field_validator("end_date")
    @classmethod
    def end_date_after_start(cls, v: date, info) -> date:
        start = info.data.get("start_date")
        if start and v <= start:
            raise ValueError("end_date must be after start_date")
        return v


class ProjectUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    budget: Optional[float] = Field(None, ge=0.1, le=99999.9)
    status: Optional[Literal["planning", "active", "completed", "on_hold"]] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None


class TaskResponse(BaseModel):
    task_id: str
    title: str
    assignee_id: str
    assignee_name: Optional[str] = None
    man_days: float
    progress: int
    start_date: date
    end_date: date
    sort_order: int
    status: Literal["not_started", "in_progress", "completed"]
    cost: Optional[float] = None


class ProjectResponse(BaseModel):
    id: str
    name: str
    description: str
    budget: float
    status: Literal["planning", "active", "completed", "on_hold"]
    start_date: date
    end_date: date
    tasks: List[TaskResponse] = []
    actual_cost: Optional[float] = None
    progress: Optional[float] = None
    created_at: datetime
    updated_at: datetime
